File: python/median.py

# Median and Sum of Absolute Deviations

# The median of a set of numbers is the value that minimizes the sum of a
# bsolute deviations from all points in the set. For a set of nn sorted 
# numbers x1,x2,…,xn the median M has the property that it 
# minimizes the sum abs(M - xi) is minimum

# given array of locations we need to find two location where 
# sum of distances from all other locations is minimum
#  
def getMinTotalDistance(dist_centers):
    # Sort the positions of the distribution centers
    dist_centers.sort()
    
    n = len(dist_centers)
    
    # Edge case: If we have fewer than two distribution centers
    if n <= 2:
        return 0
    
    # Calculate the positions for the two warehouses
    warehouse1_position = dist_centers[n // 4]
    warehouse2_position = dist_centers[3 * n // 4]

    # Calculate the total minimum distance
    total_distance = 0
    for center in dist_centers:
        total_distance += min(abs(center - warehouse1_position), abs(center - warehouse2_position))
    
    return total_distance

# using sorted list to find median of sorted array
# sorted array is like multiset balanced binary search tree
# indexing and finding element in the middle of sortedlist is o(logn) operation
from sortedcontainers import SortedList
class MedianFinder:
    def __init__(self):
        self.data = SortedList([])
    def addNum(self, num: int) -> None:
        self.data.add(num)

    def findMedian(self) -> float:
        n = len(self.data)
        if n % 2:
            return float(self.data[n // 2])
        return (float(self.data[n // 2]) + float(self.data[(n - 1) // 2])) / 2
# two heap approach, add element to max heap, the pop the max from max heap add it to mean heap
# then if length max heap is smaller than min heap pop one element from min heap and add it to max heap
#
class MedianFinder:
    def __init__(self):
        self.min_heap = []
        self.max_heap = []
        heapq.heapify(self.min_heap)
        heapq.heapify(self.max_heap)
    def addNum(self, num: int) -> None:
        heapq.heappush(self.max_heap, -num)
        heapq.heappush(self.min_heap, -heapq.heappop(self.max_heap))
        if len(self.max_heap) < len(self.min_heap):
            heapq.heappush(self.max_heap, -heapq.heappop(self.min_heap))

    def findMedian(self) -> float:
        n = len(self.min_heap) + len(self.max_heap)
        if len(self.min_heap) != len(self.max_heap):
            return float(-self.max_heap[0])
        return (float(self.min_heap[0]) + float(-self.max_heap[0])) / 2

File: Cpp/median.cpp
// here idea is to use multiset which balanced binary tree. 
// I have access to reference of each item in multiset
// now iterator should move when length of array is odd
// when size is 1 it = 1, sz = 2, it = 1, sz = 3, it = 2 ...
class MedianFinder {
public:
    std::multiset<int> data;
    std::multiset<int>::iterator it;
    MedianFinder() {

    }
    
    void addNum(int num) {
        data.insert(num);
        auto sz = data.size();
        // when length is one initialize the iterator
        if (data.size() == 1){
            it = data.begin();
        }
        // when num is smaller than it, so it goes before ir
        else if (num < *it) {
            // if size is even move back it else keep it
            it = (sz % 2 == 0) ? --it : it;
        }
        else {
            // if sz is even keep it as num is placed after it, else move it one forward
            it = (sz % 2 == 0) ? it : ++it;
        }
    }
    
    double findMedian() {
        if (data.size() % 2 == 1) {
            return static_cast<double>(*it);
        }
        auto cp = it;
        return (static_cast<double>(*cp) + static_cast<double>(*(++cp))) * 0.5;
    }
};